### PR TITLE
Introduce new singleton/class to install Python

### PR DESCRIPTION
**Pull request by nre-ableton**
The new PythonBuilder class will download a Python tarball from the
internet, extract it to the job's temporary directory, compile it, and
then return a string to the installation directory (also located in
the job's temporary directory). This takes about two minutes on one of
our single-core Linux nodes, but can be sped up with a number of
obvious optimizations.

The idea is to be able to test against new releases of python for
which packages are not yet available.

### CODE REVIEW

**Comment by nre-ableton** [comment]
Well, I didn't get any feedback on my initial API design, but I polished it up anyways and added test cases. ptal @AbletonDevTools/gotham-city, thanks!

**Comment by ala-ableton** [comment]
I am not fond of the naming. `installPython.run()` doesn't match `PythonBuilder().installI()`.

**Comment by nre-ableton** [comment]
@ala-ableton Should we go with `pythonBuilder.install()`?

**Comment by ala-ableton** [comment]
If we can, sure!

**Comment by ala-ableton** [comment]
The commit message mentions `PythonBuilderTest` instead of `PythonBuilder`.

File: src/com/ableton/PythonBuilder.groovy [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+
+class PythonBuilder implements Serializable {
+  @SuppressWarnings('FieldTypeRequired')
+  def script = null
+  /**
+   * Python major version to download (required value). Example: '3.7.0'
+   */
+  String version = null
+
```
Comment (ala-ableton) [inline src/com/ableton/PythonBuilder.groovy]: Style nitpicking, why is there an empty here?

**Comment by nre-ableton** [comment]
@ala-ableton Fixed, also I totally forgot to add the newly-created test cases to that commit (also fixed).

File: src/com/ableton/PythonBuilder.groovy [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  String revision = ''
+
+  String install() {
+    // Sanity check required values
+    assert script
+    assert version
+
+    if (!script.isUnix()) {
+      throw new UnsupportedOperationException('This function is only supported on Unix')
+    }
```
Comment (ala-ableton) [inline src/com/ableton/PythonBuilder.groovy]: What is only supported on Unix?
- Calling `download()`? Then `download()` itself should complain.
- Calling `extract()`? Then `extract()` itself should complain.
- Calling `build()`? Then `build()` itself should complain.

File: src/com/ableton/PythonBuilder.groovy (hunk 1 above)
Comment (nre-ableton) [reply src/com/ableton/PythonBuilder.groovy]: I wanted to differentiate between the required and optional parameters.

Comment (ala-ableton) [reply src/com/ableton/PythonBuilder.groovy]: I see. Then let's leave it here.

File: src/com/ableton/PythonBuilder.groovy (hunk 2 above)
Comment (nre-ableton) [reply src/com/ableton/PythonBuilder.groovy]: All of those functions will only work on Unix. 😄 This is mostly due to assumptions about path separators, but there are other things which could easily go wrong on Windows which I have not tested. For example, we call `sh` instead of `bat` which won't work on Windows hosts which lack a Unix toolchain. Our Windows build hosts have that, but we can't make any such assumptions for others (and this is a public library).

I would prefer to exclude that platform until we have a need to support it, rather than invest the time in making this code cross-platform without any real need to do so.

Am I being too lazy here? 😅

Comment (ala-ableton) [reply src/com/ableton/PythonBuilder.groovy]: You are not. I am fine with excluding Windows. But let's exclude it properly.

Comment (nre-ableton) [reply src/com/ableton/PythonBuilder.groovy]: How should we do that, then?

Comment (ala-ableton) [reply src/com/ableton/PythonBuilder.groovy]: Can we add a ctor and throw in the ctor?

Comment (nre-ableton) [reply src/com/ableton/PythonBuilder.groovy]: No, I would prefer not to do that. I have really come around to the idea of not defining ctors and letting Groovy autogenerate them, ever since we decided to go that route with `DevToolsProject`.

Let me explain the way the language works, because it might not be obvious here (it wasn't to me until I read up on it). By not defining a ctor but instead setting default values for fields, you can use "named arguments" when creating an object like so:

```groovy
class Foo {
  String bar = 'bar'
  String baz = 'baz'
}

Foo foo = new Foo(bar: 'not-bar', baz: 'not-baz')
```

I think that we can both agree that we like this syntax. 😁 So the way that works behind the scenes is that the Groovy generates a bunch of default ctors, including one which takes a `Map` and (presumably) uses all sorts of reflection to assign the values to their respective fields. The call to `new Foo` converts the args to a map behind the scenes (note that we *don't* need to use `[:]` when calling `new Foo([bar: 'not-bar' ...])`). So the point is, once you have a custom ctor, you can't create objects using the named argument syntax. In other words, this will break:

```groovy
class Foo {
  String bar
  String baz

  Foo(String bar, String baz) {
    this.bar = bar
    this.baz = baz
  }
}

// Error: Will throw groovy.lang.GroovyRuntimeException
// With a message like "Could not find matching constructor for: Foo(java.util.LinkedHashMap)"
Foo foo = new Foo(bar: 'not-bar', baz: 'not-baz')
```

So basically, named parameters aren't really named parameters. As such, I'd prefer to use the calling syntax where we can pass named parameters rather than bloat the ctor and `install` with a bunch of arguments with default values.

Comment (nre-ableton) [reply src/com/ableton/PythonBuilder.groovy]: Hmm, now that I think about it, I guess we *could* also provide a custom ctor which also takes a `Map`, but I find this code to be rather ugly:

```groovy
class Foo {
  String bar
  String baz

  Foo(Map args) {
    bar = args?.bar ?: 'bar'
    baz = args?.baz ?: 'baz'
  }
}
```

Regardless, not generating the ctor and letting Groovy do its thing is definitely the "Groovy way".

Comment (ala-ableton) [reply src/com/ableton/PythonBuilder.groovy]: Then let's not add a ctor :+1: 
I've just realized that `install` is the only public method :sweat_smile: Forget about what I said, sorry for the noise :retreat:

Comment (nre-ableton) [reply src/com/ableton/PythonBuilder.groovy]: @ala-ableton Sounds good, what changes do I need to make to this PR then? I think I've addressed your other concerns, no?

File: test/com/ableton/PythonBuilderTest.groovy [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+  }
+
+  @Test(expected = Exception)
+  void installWithBuildError() throws Exception {
+    PythonBuilder pb = new PythonBuilder(script: script, version: '1.0.0')
+    String pythonPath = pb.tempDir() + '/' + pb.pythonBaseName()
+    JenkinsMocks.addShMock('tar xfz ' + pythonPath + '.tgz', '', 0)
+    JenkinsMocks.addShMock('./configure --prefix=' + pythonPath, '', 0)
+    JenkinsMocks.addShMock('make -j1', '', 1)
+    JenkinsMocks.addShMock('make install', '', 0)
```
Comment (ala-ableton) [inline test/com/ableton/PythonBuilderTest.groovy]: Is it really necessary to add that sh mock for `make install`?

File: test/com/ableton/PythonBuilderTest.groovy [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+  @Test(expected = IllegalArgumentException)
+  void installWithEmptyMakeJobs() throws Exception {
+    PythonBuilder pb = new PythonBuilder(script: script, version: '1', makeJobs: '')
+    String pythonPath = pb.tempDir() + '/' + pb.pythonBaseName()
+    JenkinsMocks.addShMock('tar xfz ' + pythonPath + '.tgz', '', 0)
+    pb.install()
+  }
+
+  @Test(expected = NumberFormatException)
```
Comment (ala-ableton) [inline test/com/ableton/PythonBuilderTest.groovy]: What is throwing this exception?

File: test/com/ableton/PythonBuilderTest.groovy (hunk 3 above)
Comment (nre-ableton) [reply test/com/ableton/PythonBuilderTest.groovy]: No, it's not, so I will remove this. FYI this was necessary with jenkins-pipeline-mocks 0.2.0 (which I started out with when I wrote these tests), but in 0.4.1 we fixed `sh`'s exception throwing behavior so it is no longer necessary.

File: test/com/ableton/PythonBuilderTest.groovy (hunk 4 above)
Comment (nre-ableton) [reply test/com/ableton/PythonBuilderTest.groovy]: The `toInteger()` call in `build()`.

**Comment by nre-ableton** [comment]
FYI @ala-ableton, @rco-ableton, here's a preview of what mergeomatic's Jenkinsfile will look like with this: https://github.com/AbletonDevTools/mergeomatic/pull/74

It fails on Jenkins because we will need some extra software installed on our Linux nodes to build Python 3.7, namely `zlib1-dev` and a few other packages which currently escape me. 😅

File: test/com/ableton/PythonBuilderTest.groovy (hunk 4 above)
Comment (ala-ableton) [reply test/com/ableton/PythonBuilderTest.groovy]: I see.
